weaviate/event_loop.py

```diff
@@ -0,0 +1,97 @@
+import asyncio
+import threading
+import time
+from concurrent.futures import Future
+from typing import Any, Callable, Coroutine, Generic, Optional, TypeVar, cast
+
+from typing_extensions import ParamSpec
+
+from weaviate.exceptions import WeaviateClosedClientError
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+class _Future(Future, Generic[T]):
+    def result(self, timeout: Optional[float] = None) -> T:
+        return cast(T, super().result(timeout))
+
+
+class _EventLoop:
+    def __init__(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+        self.loop = loop
+
+    def start(self) -> None:
+        if self.loop is not None:
+            return
+        self.loop = self.__start_new_event_loop()
+
+    def run_until_complete(
+        self, f: Callable[P, Coroutine[Any, Any, T]], *args: P.args, **kwargs: P.kwargs
+    ) -> T:
+        """This method runs the provided coroutine in a blocking manner by scheduling its execution
+        in an event loop running in a parallel thread.
+
+        The result of the coroutine is returned, either when the coroutine completes or raises an exception.
+        """
+        if self.loop is None or self.loop.is_closed():
+            raise WeaviateClosedClientError()
+        fut = asyncio.run_coroutine_threadsafe(f(*args, **kwargs), self.loop)
+        return fut.result()
+
+    def schedule(
```

Comment:
does not seem to be used?

Reply:
It's used on `L300` of `weaviate.collections.batch.base.py`

Reply:
the github UI didn't show that :)

Reply:
Probably overloaded with the number of changed lines 😅